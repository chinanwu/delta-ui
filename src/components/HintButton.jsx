import PropTypes from 'prop-types';
import React, { useCallback } from 'react';
import { connect } from 'react-redux';

import getThemeClassname from '../functions/getThemeClassname';

import './HintButton.less';

export const HintButton = ({
	id,
	dark,
	hintWord,
	hintNumLeft,
	numHints,
	isExpanded,
	giveSolution,
	onSolnClick,
	onHint,
	onClose,
}) => {
	const handleClick = useCallback(
		() =>
			isExpanded
				? onClose()
				: numHints > 0
				? onHint()
				: giveSolution
				? onSolnClick()
				: null,
		[isExpanded, numHints, giveSolution, onHint, onSolnClick]
	);

	return (
		<div
			className={
				getThemeClassname('HintButton', dark) +
				(isExpanded ? ' HintButton--expanded' : '')
			}
		>
			<div
				className={
					getThemeClassname('HintButton__content', dark) +
					(isExpanded ? ' HintButton__content--expanded' : '')
				}
				aria-labelledby="hintHeader"
				aria-live="passive"
			>
				<h3 id="hintHeader">Hint: </h3>
				<p className="HintButton__hint">Recommended next word: {hintWord}</p>
				<p className="HintButton__hint">
					Estimated number of words left: {hintNumLeft}
				</p>
			</div>
			<button
				id={id}
				className={
					'HintButton__btn' +
					(isExpanded
						? ' HintButton__btn--expanded'
						: !giveSolution && numHints === 0 && dark
						? ' HintButton__btn--disabled--dark'
						: '')
				}
				aria-label={
					isExpanded
						? 'Close Hint'
						: numHints === 0
						? giveSolution
							? 'Get the solution'
							: 'Out of Hints'
						: 'Get a Hint'
				}
				aria-expanded={isExpanded}
				aria-haspopup={!isExpanded && numHints === 0 ? 'dialog' : null}
				disabled={!giveSolution && numHints === 0 && !isExpanded}
				onClick={handleClick}
			>
				{isExpanded
					? 'Close Hint'
					: numHints === 0
					? giveSolution
						? 'Get the solution'
						: 'Out of Hints'
					: `Get a Hint: ${numHints}`}
			</button>
		</div>
	);
};

export const mapStateToProps = ({ theme: { dark } }) => ({
	dark,
});

HintButton.propTypes = {
	id: PropTypes.string,
	dark: PropTypes.bool,
	hintWord: PropTypes.string,
	hintNumLeft: PropTypes.number,
	numHints: PropTypes.number,
	isExpanded: PropTypes.bool,
	giveSolution: PropTypes.bool,
	onHint: PropTypes.func,
	onSolnClick: PropTypes.func,
	onClose: PropTypes.func,
};

export default connect(mapStateToProps)(HintButton);
