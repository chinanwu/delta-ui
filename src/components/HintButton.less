@import '../style/button';
@import '../style/colours';

.HintButton {
	width: 30rem;
	height: 17rem;
}

.HintButton__content {
	background-color: @greenLight;
	height: 0;
	transition: height 1s;
	overflow: hidden;
	margin-left: 1rem;
	margin-right: 1rem;
}

.HintButton__content--expanded {
	height: 13rem;
}

.HintButton__content--dark {
	background-color: @dark-02;
	color: #fff;
}

.HintButton__btn {
	@medBtn();
	margin-top: 0;
}

.HintButton__btn--expanded {
	border-radius: 0 0 0.5rem 0.5rem;
}

.HintButton__btn--disabled--dark {
	@btnDark();
}

@media only screen and (max-width: 550px) {
	.HintButton {
		width: initial;
		height: initial;
		display: flex;
		flex-direction: column-reverse;
		align-items: center;
	}

	.HintButton--expanded {
		align-items: initial;
	}

	.HintButton__btn {
		margin-top: 1rem;
		margin-bottom: 0;
		margin-right: 2rem;
		width: 20rem;
	}

	.HintButton__btn--expanded {
		border-radius: 0.5rem 0.5rem 0 0;
		width: 13rem;
	}
}
