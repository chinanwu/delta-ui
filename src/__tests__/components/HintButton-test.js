import { shallow } from 'enzyme';
import React from 'react';

import { HintButton } from '../../components/HintButton.jsx';

jest.unmock('../../components/HintButton.jsx');

describe('HintButton', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(<HintButton id="test" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders HintButton dark', () => {
			const wrapper = shallow(<HintButton id="test" dark={true} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders HintButton light', () => {
			const wrapper = shallow(<HintButton id="test" dark={false} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expanded HintButton', () => {
			const wrapper = shallow(<HintButton id="test" isExpanded={true} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {});
});
