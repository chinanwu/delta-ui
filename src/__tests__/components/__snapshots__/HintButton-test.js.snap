// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HintButton behaviour closes on click when expanded 1`] = `
<div
  className="HintButton HintButton--expanded"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content HintButton__content--expanded"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-expanded={true}
    aria-haspopup={null}
    aria-label="Close Hint"
    className="HintButton__btn HintButton__btn--expanded"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Close Hint
  </button>
</div>
`;

exports[`HintButton behaviour does nothing on click when out of hints and not give solution 1`] = `
<div
  className="HintButton"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-expanded={false}
    aria-haspopup="dialog"
    aria-label="Out of Hints"
    className="HintButton__btn"
    disabled={true}
    id="test"
    onClick={[Function]}
  >
    Out of Hints
  </button>
</div>
`;

exports[`HintButton behaviour gets hint on click when closed 1`] = `
<div
  className="HintButton"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-expanded={false}
    aria-haspopup={null}
    aria-label="Get a Hint"
    className="HintButton__btn"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Get a Hint: 3
  </button>
</div>
`;

exports[`HintButton behaviour gets solution on click when out of hints 1`] = `
<div
  className="HintButton"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-expanded={false}
    aria-haspopup="dialog"
    aria-label="Get the solution"
    className="HintButton__btn"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Get the solution
  </button>
</div>
`;

exports[`HintButton rendering renders HintButton dark 1`] = `
<div
  className="HintButton HintButton--dark"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content HintButton__content--dark"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-haspopup={null}
    aria-label="Get a Hint"
    className="HintButton__btn"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Get a Hint: undefined
  </button>
</div>
`;

exports[`HintButton rendering renders HintButton light 1`] = `
<div
  className="HintButton"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-haspopup={null}
    aria-label="Get a Hint"
    className="HintButton__btn"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Get a Hint: undefined
  </button>
</div>
`;

exports[`HintButton rendering renders expanded HintButton 1`] = `
<div
  className="HintButton HintButton--expanded"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content HintButton__content--expanded"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-expanded={true}
    aria-haspopup={null}
    aria-label="Close Hint"
    className="HintButton__btn HintButton__btn--expanded"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Close Hint
  </button>
</div>
`;

exports[`HintButton rendering renders expected elements 1`] = `
<div
  className="HintButton"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-haspopup={null}
    aria-label="Get a Hint"
    className="HintButton__btn"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Get a Hint: undefined
  </button>
</div>
`;

exports[`HintButton rendering renders no hints left for daily 1`] = `
<div
  className="HintButton HintButton--dark"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content HintButton__content--dark"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-haspopup="dialog"
    aria-label="Out of Hints"
    className="HintButton__btn HintButton__btn--disabled--dark"
    disabled={true}
    id="test"
    onClick={[Function]}
  >
    Out of Hints
  </button>
</div>
`;

exports[`HintButton rendering renders no hints left for solo 1`] = `
<div
  className="HintButton HintButton--dark"
>
  <div
    aria-labelledby="hintHeader"
    aria-live="passive"
    className="HintButton__content HintButton__content--dark"
  >
    <h3
      id="hintHeader"
    >
      Hint: 
    </h3>
    <p
      className="HintButton__hint"
    >
      Recommended next word: 
    </p>
    <p
      className="HintButton__hint"
    >
      Estimated number of words left: 
    </p>
  </div>
  <button
    aria-haspopup="dialog"
    aria-label="Get the solution"
    className="HintButton__btn"
    disabled={false}
    id="test"
    onClick={[Function]}
  >
    Get the solution
  </button>
</div>
`;
